cip/1.accepted/CIP2017-02-06-Regular-Path-Patterns.adoc

```diff
@@ -0,0 +1,160 @@
+
+----
+PatternElementChain = (RelationshipPattern | RegularPathPattern), NodePattern ;
+
+RegularPathPattern = (LeftArrowHead, Dash, '/', [RegularPathExpression], '/', Dash, RightArrowHead)
+                   | (LeftArrowHead, Dash, '/', [RegularPathExpression], '/', Dash)
+                   | (Dash, '/', [RegularPathExpression], '/', Dash, RightArrowHead)
+                   | (Dash, '/', [RegularPathExpression], '/', Dash)
+                   ;
+RegularPathExpression = {RegPathOr}- ;
+RegPathOr = RegPathSeq, {'|', RegPathSeq} ;
+RegPathSeq = {RegPathStar}- ;
+RegPathStar = RegPathDirected [('*', [RangeLiteral]) | '+'] ;
+RegPathDirected = ['<'], RegPathBase, ['>'] ;
+RegPathBase = RegPathRelationship
+            | RegPathReference
+            | '(' RegularPathExpression ')'
+            ;
+RegPathRelationship = RelType ;
+RegPathReference = SymbolicName ;
+----
+
+The `RegPathReference` is a reference to a Defined Path Predicate.
+These are defined using the following syntax:
+
+----
+DefinedPathPredicate = PathPredicatePrototype, 'IS', Pattern, [Where] ;
+PathPredicatePrototype = '(', Variable, ')', RegPathPrototype, '(', Variable, ')' ;
+RegPathPrototype = (LeftArrowHead, Dash, '/', DefinedPathName, '/', Dash)
+                   | (Dash, '/', DefinedPathName, '/', Dash, RightArrowHead)
+                   | (Dash, '/', DefinedPathName, '/', Dash)
+                   ;
+DefinedPathName = SymbolicName ;
+----
+
+=== Examples
+
+The astute reader of the syntax will have noticed that it is possible to express a Regular Path Pattern with an empty path expression:
+
+[source, cypher]
+----
+MATCH (a)-//-(b)
+----
+
+This pattern simply states that `a` and `b` must be the same node, and is thus the same as:
+
+[source, cypher]
+----
+MATCH (a), (b) WHERE a = b
+----
+
+The same reader will also have noticed that it is possible to define a pattern containing just a relationship type:
+
+[source, cypher]
+----
+MATCH (a)-/:KNOWS/->(b)
+----
+
+That pattern is indeed equivalent to the very similar relationship pattern:
+
+[source, cypher]
+----
+MATCH (a)-[:KNOWS]->(b)
+----
+
+The main difference being that the variant with a relationship pattern is able to bind that relationship and express further predicates over it.
+
+The Regular Path Patterns start becoming interesting when larger expressions are put together:
+
+[source, cypher]
+.Finding someone loved by someone hated by someone you know, transitively
+----
+MATCH (you)-/(:KNOWS :HATES)+ :LOVES/->(someone)
+----
+
+Note the `+` expressing one or more occurrences of the sequence `KNOWS` followed by `HATES`.
+
+The direction of each relationship is governed by the overall direction of the Regular Path Pattern.
+It is however possible to explicitly define the direction for a particular part of the pattern.
+This is done by either prefixing that part with `<` for a right-to-left direction or suffixing it with `>` for a left-to-right direction.
```

Comment:
We could use 'outgoing' instead of 'left-to-right', and 'incoming' instead of 'right-to-left'.